composer/utils/fx_utils.py

```diff
@@ -123,16 +125,136 @@ def detect_residual_pattern(gm: GraphModule):
+            assert len(node.all_input_nodes) == 2
+            node0, node1 = node.all_input_nodes[0], node.all_input_nodes[1]
+            lhs_nodes, rhs_nodes = _get_residual_block_nodes(node0, node1)
+            if lhs_nodes or rhs_nodes:
+                if len(lhs_nodes):
+                    _attach_tag(lhs_nodes, f'non_res_{count}')
+                    all_tags.append(f'non_res_{count}')
+                if len(rhs_nodes):
+                    _attach_tag(rhs_nodes, f'residual_{count}')
+                    all_tags.append(f'residual_{count}')
+                add_tag = f'addN_{count}'
+                if add_tag not in all_tags:
+                    all_tags.append(add_tag)
+                node.tag = add_tag
+                count += 1
+    return all_tags, count
+
+
+def _get_residual_modules(gm: GraphModule, node: Node) -> Tuple[Optional[GraphModule], Optional[GraphModule], int]:
+    """Returns GraphModules for the main and residual branches.
+
+    node.op is assumed to be a call_module
+    """
+    pattern = re.compile(r'non_res_(\d+)|residual_(\d+)')
+    matches = pattern.match(str(node.target))
+    if matches:
+        idx = int(matches[1]) if matches[1] else int(matches[2])
+        main_submod = getattr(gm, f'non_res_{idx}')
+        residual_submod = getattr(gm, f'residual_{idx}', None)
+        return main_submod, residual_submod, idx
+    else:
+        return None, None, 0
+
+
+def _replace_residual_pattern(gm: GraphModule,
+                              original_node: Node,
+                              replacement_module: str,
+                              has_residual_ops: bool = False) -> None:
+    """Replaces main, residual and add_node with the ``replacement_module``.
+
+    ``replacement_module`` is already added to the gm.
+    """
+    insert_node = original_node.prev
+    add_node = original_node.next
+    if has_residual_ops:
+        add_node = original_node.next.next
+    with gm.graph.inserting_after(insert_node):
+        new_node = gm.graph.call_module(replacement_module, args=(insert_node,))  # type: ignore
+        add_node.replace_all_uses_with(new_node)
+        gm.graph.erase_node(add_node)
+        if has_residual_ops:
+            gm.graph.erase_node(original_node.next)
+        gm.graph.erase_node(original_node)
+    gm.graph.lint()
+
+
+def apply_stochastic_residual(gm: GraphModule, drop_rate: float = 0.2) -> Tuple[GraphModule, int]:
+    """Detect and replace residual pattern with their stochastic equivalent.
+
+    Arguments:
+        gm (GraphModule): The source FX-traced graph. It can be the whole model symbolically traced.
+
+    Returns:
+        GraphModule: Modified GraphModule that has stochastic residual connections.
     """
-    raise NotImplementedError('replace_residual_with_stochastic is currently not implemented.')
+    assert isinstance(gm, GraphModule), 'Input to apply_stochastic_residual should be an instance of GraphModule'
```

Comment:
since `apply_*` is part of the public API, this should raise a `ValueError` instead of an assert.

Reply:
Converted it to valueerror